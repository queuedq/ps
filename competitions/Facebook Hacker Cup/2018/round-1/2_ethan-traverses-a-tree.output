Case #1: 1 1 2 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 2 2 2 2 1 3 4 1 1 2 1 2 2 1 1 1 1 1 2 2 2 1 2 4 2 1 1 1 1 3 1 2 5 1 1 1 2 2 2 1 1 1 1 1 1 1 1 6 1 2 1 1 1 5 1 3 1 1 1 1 2 2 1 7 1 4 7 1 2 1 1 1 1 1 1 2 2 2 2 1 1 1 8 2 2 1 1 1 2 1 1 1 2 1 9 1 1 2 1 2 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 2 2 1 1 1 1 1 1 10 1 1 2 1 1 1 11 1 1 1 2 12 2 2 13 1 1 1 1 4 1 1 2 1 1 14 2 1 1 2 2 1 1 1 3 1 15 16 2 1 2 1 1 17 18 2 1 19 2 1 1 2 2 1 1 1 1 19 1 1 1 1 19 2 12 1 1 2 17 2 1 1 1 1 2 2 1 1 2 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 10 1 1 2 1 2 1 1 19 2 2 1 1 1 1 1 1 1 19 1 1 1 2 19 1 1 2 1 1 19 1 1 1 18 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 15 1 1 2 2 1 1 2 2 1 1 1 1 1 19 1 1 13 2 1 1 1 1 1 1 1 1 1 2 1 1 1 2 2 1 1 1 1 1 1 2 1 1 1 1 1 2 2 1 1 1 1 1 1 1 2 1 1 1 3 1 1 4 2 19 1 1 11 1 2 1 1 2 1 1 2 2 1 1 1 1 1 2 2 1 1 1 2 2 1 1 1 1 1 1 2 1 1 1 1 1 1 2 1 1 2 2 1 1 2 1 2 2 1 1 2 1 2 1 2 1 1 1 2 2 2 1 2 1 1 1 1 1 1 1 1 1 1 2 2 1 1 4 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 1 2 3 2 1 1 2 1 1 1 1 2 2 1 1 1 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 1 2 1 2 2 1 1 1 2 2 1 1 1 1 1 1 2 1 1 1 2 1 1 1 1 1 1 1 1 1 2 1 1 2 1
Case #2: 1 1 2 1 2 1 1 2 2 2 1 1 2 1 1 2 2 1 1 1 3 1 1 1 2 4 2 1 5 2 1 1 1 2 2 1 2 2 1 1 2 1 1 1 1 5 1 1 1 4 1 2 1 6 1 1 2 7 1 2 2 2 1 1 1 1 2 1 2 2 1 8 7 7 1 2 4 2 1 1 2 1 7 1 1 9 1 1 1 7 1 2 1 2 2 1 2 4 1 1 1 10 1 1 1 1 1 1 1 1 1 11 12 5 2 1 1 2 1 1 1 2 13 1 2 2 14 1 15 5 16 2 1 1 1 2 1 1 1 1 17 1 15 2 18 1 1 1 1 1 19 1 2 20 21 1 1 2 1 1 21 1 2 19 2 2 21 2 2 1 16 1 1 2 2 4 2 5 7 2 1 21 2 6 20 1 1 1 5 2 4 17 1 1 2 1 2 2 1 2 7 1 21 1 1 2 1 1 1 2 1 21 2 21 1 1 1 21 7 5 2 2 5 1 1 5 1 5 1 1 21 1 2 3 2 1 1 4 21 1 21 1 1 1 1 1 21 21 2 1 1 1 1 1 1 21 21 1 2 1 1 2 5 21 1 7 2 1 1 2 1 2 2 3 1 3 7 21 2 1 1 2 2 7 1 1 21 2 1 5 21 1 21 2 1 1 1 21 21 7 1 2 1 2 1 21 15 21 2 2 1 1 2 15 5 1 1 1 2 2 21 6 1 2 21 10 1 2 7 6 9 7 1 1 1 2 1 1 3 1 7 1 2 21 1 1 1 21 21 2 1 2 1 1 2 2 1 1 1 2 21 7 21 1 1 2 1 2 6 2 2 2 7 2 1 1 6 7 21 21 2 2 2 1 2 2 2 2 1 21 1 1 21 7 7 1 1 1 7 1 21 7 1 1 21 1 1 1 2 21 1 1 21 2 1 2 1 21 1 1 1 21 4 21 2 1 7 2 21 2 2 1 1 1 21 1 1 1 2 1 1 1 1 1 7 1 1 2 1 1 1 1 1 2 2 7 1 1 21 1 2 21 1 1 1 4 2 21 2 1 1 1 7 2 1 3 2 1 1 1 1 1 21 1 4 2 2 1 1 7 1 1 13 2 21 1 1 1 2 1 2 5 2 2 1 7 1 2 2 1 7 2 1 1 1 1 7 1 1 1 1 2 2 1 1 1 1 1 21 21 21 1 2 2 21 2 21 1 2 2 21 2 1 2 21 1 1 2 2 2 9 3 2 10 2 5 2 5 2 1 7 5 4 1 4 7 1 10 4 2 1 1 4 1 1 2 2 1 21 1 1 2 21 5 21 18 2 21 2 2 1 1 21 2 2 2 7 2 2 1 2 2 1 1 1 1 6 2 2 2 2 2 2 1 1 2 7 1 21 2 1 16 1 2 1 1 1 21 2 2 5 1 5 2 2 3 21 1 2 21 1 2 2 1 1 2 1 1 1 1 21 1 21 1 2 1 1 2 1 2 1 1 7 1 1 2 15 1 2 2 1 2 2 1 1 21 18 2 4 1 1 1 2 1 7 1 1 1 1 3 1 1 1 2 1 1 1 1 1 1 1 1 1 1 2 2 7 2 1 1 1 13 1 1 2 4 1 12 1 1 1 1 2 1 2 1 1 2 1 1 1 1 2 1 1 3 1 1 1 1 1 1 21 2 7 1 2 7 15 21 2 21 16 1 1 2 2 21 2 1 1 7 2 1 1 1 2 1 1 2 3 2 1 1 1 1 2 1 1 2 1 1 9 7 21 2 1 2 2 1 2 7 1 1 1 2 2 1 1 1 2 1 18 7 2 2 2 1 6 6 1 1 2 2 2 1 2 21 1 1 1 1 2 2 1 6 7 2 2 2 1 5 7 2 1 2 2 7 4 1 1 1 1 6 2 2 1 1 1 1 2 1 1 1 7 3 1 1 21 2 2 1 1 2 1 1 1 21 1 1 1 1 2 1 1 1 7 7 2 1 21 2 7 1 1 1 2 2 1 10 7 5 2 1 2 4 2 1 1 2 2 1 1 2 1 2 2 1 1 7 1 4 1 2 1 2 21 21 1 1 7 1 2 18 1 1 1 1 1 4 2 4 1 2 7 1 1 15 2 1 1 2 1 1 1 2 1 1 1 2 1 1 1 15 2 1 5 1 1 1 5 1 1 1 1 1 1 1 1 7 2 1 1 2 2 1 2 1 1 1 1 12 1 2 1 1 2 1 10 1 1 1 2 1 1 1 7 1 1 7 2 2 2 1 2 2 21 1 2 2 1 1 2 1 7 1 2 2 1 16 2 1 1 1 2 2 1 21 1 2 1 5 1 2 6 2 1 1 18 21 1 21 2 2 1 5 13 1 1 1 1 1 1 1 1 1 1 2 2 2 1 1 2 2 1 2 1 2 16 2 1 1 2 1 2 1 1 2 2 5 15 1 7 1 2 16 2 1 21 4 5 2 2 6 1 2 1 1 1 7 1 1 2 2 6 1 2 1 1 1 1 21 1 2 7 1 10 1 1 1 2 2 2 7 1 1 1 2 2 1 2 21 9 2 7 1 4 2 7 1 2 2 2 2 7 1 5 1 1 1 2 5 2 1 18 1 1 1 1 2 2 1 1 1 1 1 1 1 1 5 1 2 1 1 1 1 1 7 7 1 1 1 1 7 15 1 1 5 1 7 2 11 1 1 1 2 1 1 1 1 1 1 1 1 1 1 1 2 1 2 1 7 1 1 1 1 2 2 1 2 1 2 1 1 1 1 1 1 21 3 1 1 1 21 1 4 1 6 1 2 1 2 1 1 2 2 2 2 1 5 2 1 1 1 1 1 1 1 2 1 1 1 15 2 2 1 1 2 1 2 1 1 1 6 1 1 2 21 2 1 1 2 2 1 1 1 1 21 1 1 1 1 2 1 1 1 2 1 2 1 1 1 2 2 1 1 1 2 1 1 2 1 2 4 2 1 2 5 1 7 2 1 3 1 2 3 1 7 1 2 1 2 1 1 7 2 1 1 1 1 1 1 1 1 1 1 13 5 2 5 2 2 2 2 1 1 2 1 1 1 1 1 1 1 1 1 15 2 2 1 2 2 2 1 1 1 1 7 7 1 1 2 1 21 7 1 1 1 1 1 1 2 16 1 1 7 2 2 7 2 1 1 7 2 1 1 1 18 5 3 2 1 1 2 4 2 2 1 1 1 2 1 1 2 1 1 1 1 1 1 1 1 1 2 21 4 1 1 1 1 1 1 2 2 2 1 1 1 1 1 2 1 14 1 1 2 2 1 2 1 1 1 1 7 1 1 7 1 2 1 1 1 1 1 1 4 1 1 2 2 2 1 2 7 1 4 1 1 1 1 1 4 6 1 7 1 1 1 1 4 1 2 1 1 1 2 1 1 1 2 1 2 1 1 1 1 1 2 5 2 10 5 1 1 1 2 1 2 1 1 1 1 1 2 1 1 2 1 1 7 4 1 1 2 5 1 2 1 1 1 2 2 1 1 5 2 1 1 2 2 1 7 1 21 2 1 1 2 2 1 1 1 16 1 1 7 1 1 1 7 2 2 5 1 1 2 1 5 1 2 2 1 7 1 2 1 2 1 2 2 1 2 11 1 2 2 1 2 2 2 4 21 7 1 1 2 2 7 1 7 1 2 1 1 1 1 1 7 1 2 1 6 10 1 1 1 7 7 1 2 1 1 14 1 1 2 1 2 2 2 1 4 1 1 1 6 7 1 1 1 1 7 2 4 4 18 1 2 1 1 2 2 1 1 1 2 1 5 7 1 1 1 2 1 1 7 7 2 2 1 7 1 1 1 2 1 1 2 2 1 1 2 1 1 1 1 7 2 2 1 1 1 7 21 1 1 1 1 1 3 7 1 5 1 5 2 7 1 2 1 1 15 1 2 2 2 1 1 2 2 2 2 1 1 2 1 1 1 1 4 18 2 1 2 5 2 1 2 1 1 2 7 2 2 2 1 2 2 1 2 2 1 8 1 1 1 2 1 1 1 1 21 2 1 2 2 12 1 1 2 1 2 1 2 1 2 2 21 2 1 1 1 2 7 2 2 1 5 1 1 2 2 1 2 1 1 2 1 1 1 1 2 1 1 1 18 1 2 2 1 1 1 2 1 7 2 1 2 7 1 1 3 2 2 2 1 2 2 1 1 1 2 1 7 1 1 2 2 21 1 1 1 1 1 1 1 1 1 1 2 4 1 7 2 1 4 1 1 7 1 2 2 1 1 1 1 2 2 2 2 1 1 1 2 1 2 1 15 1 11 1 1 4 7 1 1 1 2 1 7 2 1 4 5 1 1 2 1 2 2 1 1 1 1 1 1 1 2 1 1 2 1 2 1 1 1 1 1 2 2 2 1 1 1 4 2 1 2 7 1 1 2 2 1 2 2 1 1 2 1 4 2 1 1 1 2 1 1 1 7 1 1 7 1 2 2 1 1 1 11 2 1 5 7 1 1 1 2 1 1 2 1 2 1 2 1 1 1 1 1
Case #3: Impossible
Case #4: 1 2 2 2 2 2 1 1 1 1 2 1 2 1 3 1 1 1 2 1 1 3 2 1 4 1 1 1 1 2 1 5 2 1 5 1 6 1 1 2 2 2 1 2 1 2 1 1 1 1 5 2 2 1 2 1 4 6 1 1 7 6 2 1 1 1 2 1 2 2 4 8 2 2 6 4 1 1 1 2 1 1 1 1 1 1 1 1 1 2 2 1 2 1 1 1 1 1 2 1 8 8 1 1 1 2 1 4 2 1 8 1 1 2 1 2 2 1 2 1 1 1 1 1 1 1 2 1 1 1 1 1 1 1 1 1 6 1 2 2 1 1 8 1 1 2 1 1 2 1 1 1 1 1 4 1 6 2 8 1 1 2 2 2 5 2 1 8 1 1 1 1 6 2 1 1 1 1 1 1 2 1 1 1 6 2 1 1 1 1 1 6 2 2 1 6 2 1 1 2 1 1 1 1 2 6 8 8 1 2 1 1 2 2 5 1 8 1 1 2 1 1 2 8 2 2 1 1 1 1 2 1 2 1 1 6 1 2 1 1 2 1 1 1 2 2 6 1 4 1 2 1 1 1 8 1 2 1 1 5 8 1 1 8 1 8 1 1 2 2 5
Case #5: 1 1 1 2 1 1 2 1 2 1 2 1 2 1 2 2 1 3 1 4 5 1 6 7 1 2 8 9 1 10 11 1 1 12 2 2 1 13 14 2 2 15 1 16 17 18 19 1 1 20 21 20 19 18 17 16 2 2 15 14 1 13 12 1 2 11 2 2 2 10 9 1 2 8 1 1 2 7 6 1 1 2 2 1 5 2 1 22 2 1 2 2 4 2 1 1 3 1 2 1 1 1 2 1 1 2 1 2 1 1 1 1 2 2 22 1 2 1 2 2 1 1 1 1 2 2 1 1 2 2 1 1 2 22 1 1 1 1 1 1 1 1 1 1 2 1 1 1 2 2 2 1 2 1 22 1 1 2 1 2 1 1 1 1 1 1
Case #6: Impossible
