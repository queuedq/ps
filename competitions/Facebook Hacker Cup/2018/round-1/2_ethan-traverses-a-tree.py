class DisjointSet:
  def __init__(self, totalSize):
    self.parent = list(range(totalSize))
    self.size = [1] * totalSize

  def find(self, x):
    while x != self.parent[x]:
      x = self.parent[x]
      self.parent[x] = self.parent[self.parent[x]]
    return x

  def union(self, x, y):
    xRoot = self.find(x)
    yRoot = self.find(y)

    if self.size[xRoot] < self.size[yRoot]:
      xRoot, yRoot = yRoot, xRoot

    self.parent[yRoot] = xRoot
    self.size[xRoot] = self.size[xRoot] + self.size[yRoot]

  def __str__(self):
    return str(self.parent)

def traverse(node, left, right, pre, post):
  if node == 0:
    return
  pre.append(node)
  traverse(left[node], left, right, pre, post)
  traverse(right[node], left, right, pre, post)
  post.append(node)


def ethan():
  N, K = (int(x) for x in input().split())
  left, right = [0], [0]
  for node in range(N):
    l, r = (int(x) for x in input().split())
    left.append(l)
    right.append(r)

  pre, post = [], []

  traverse(1, left, right, pre, post)

  ## print(N, K)
  ## print(pre)
  ## print(post)

  labels = DisjointSet(N+1)
  for i in range(N):
    labels.union(pre[i], post[i])

  labelMapping = [None] * (N+1)
  cnt = 0
  for i in range(1, N+1):
    root = labels.find(i)
    if labelMapping[root] is None:
      cnt += 1
      if cnt > K:
        cnt = K
      labelMapping[root] = cnt
      labelMapping[i] = cnt
    else:
      labelMapping[i] = labelMapping[root]

  if cnt < K:
    return 'Impossible'
  else:
    return ' '.join([str(x) for x in labelMapping[1:]])


T = int(input())

for test in range(1, T+1):
  print('Case #{}: {}'.format(test, ethan()))
