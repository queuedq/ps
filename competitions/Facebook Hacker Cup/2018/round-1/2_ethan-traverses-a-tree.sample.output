Case #1: 1 1
Case #2: Impossible
Case #3: 1 1 2
Case #4: 1 2 3 4 4 2 3 1 1
Case #5: 1 1 2 3 4 1 2 3 1 3 1 3 2 3 1
Case #6: Impossible
