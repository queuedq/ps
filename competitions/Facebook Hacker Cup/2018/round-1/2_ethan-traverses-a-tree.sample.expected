Case #1: 1 1
Case #2: Impossible
Case #3: 2 2 1
Case #4: 2 4 3 1 3 4 3 2 2
Case #5: 1 1 2 3 4 1 2 3 1 3 1 3 2 3 1
Case #6: Impossible
